File: core/src/net/natruid/jungle/systems/ThreatSystem.java
package net.natruid.jungle.systems;

import com.artemis.Aspect;
import com.artemis.BaseEntitySystem;
import com.artemis.ComponentMapper;
import com.artemis.utils.IntBag;
import net.natruid.jungle.components.AttributesComponent;
import net.natruid.jungle.components.BehaviorComponent;
import net.natruid.jungle.components.UnitComponent;
import net.natruid.jungle.utils.types.AttributeType;

public class ThreatSystem extends BaseEntitySystem {
    private ComponentMapper<BehaviorComponent> mBehavior;
    private ComponentMapper<UnitComponent> mUnit;
    private ComponentMapper<AttributesComponent> mAttributes;
    private UnitManageSystem unitManageSystem;
    private TileSystem tileSystem;

    public ThreatSystem() {
        super(Aspect.all(BehaviorComponent.class));
    }

    public void checkAlert(int unit) {
        checkAlert(unit, -1);
    }

    public void checkAlert(int unit, int tile) {
        IntBag entityIds = getEntityIds();
        if (entityIds.isEmpty()) return;
        if (!mUnit.has(unit)) return;
        int t = (tile >= 0) ? tile : mUnit.get(unit).tile;
        int[] data = entityIds.getData();
        for (int i = 0; i < entityIds.size(); i++) {
            int u = data[i];
            if (!unitManageSystem.isEnemy(u, unit)) continue;
            if (!mBehavior.get(u).threatMap.isEmpty()) continue;
            float maxDistance = 6f * 1f + (mAttributes.get(u).get(AttributeType.AWARENESS) - 10) * 0.05f;
            if (tileSystem.getDistance(mUnit.get(u).tile, t) <= maxDistance) {
                mBehavior.get(u).threatMap.put(u, 1f);
            }
        }
    }

    @Override
    protected void processSystem() {
    }
}

File: core/src/net/natruid/jungle/systems/ThreatSystem.kt
package net.natruid.jungle.systems

import com.artemis.Aspect
import com.artemis.BaseEntitySystem
import com.artemis.ComponentMapper
import net.natruid.jungle.components.AttributesComponent
import net.natruid.jungle.components.BehaviorComponent
import net.natruid.jungle.components.UnitComponent

class ThreatSystem : BaseEntitySystem(Aspect.all(BehaviorComponent::class.java)) {
    private lateinit var mBehavior: ComponentMapper<BehaviorComponent>
    private lateinit var mUnit: ComponentMapper<UnitComponent>
    private lateinit var mAttributes: ComponentMapper<AttributesComponent>
    private lateinit var unitManageSystem: UnitManageSystem
    private lateinit var tileSystem: TileSystem

    fun checkAlert(unit: Int, tile: Int = -1) {
        if (entityIds.isEmpty) return
        if (!mUnit.has(unit)) return
        val t = if (tile >= 0) tile else mUnit[unit].tile
        val data = entityIds.data
        for (i in 0 until entityIds.size()) {
            val u = data[i]
            if (!unitManageSystem.isEnemy(u, unit)) continue
            if (mBehavior[u].threatMap.isNotEmpty()) continue
            val maxDistance = 6f * 1f + (mAttributes[u].awareness - 10) * 0.05f
            if (tileSystem.getDistance(mUnit[u].tile, t) <= maxDistance) {
                mBehavior[u].threatMap[u] = 1f
            }
        }
    }

    override fun processSystem() {}
}
